### DISCUSSION TITLE
Colmap2nerf script transformation

### DISCUSSION THREAD

**Discussion by saimouli** [NVlabs/instant-ngp#1319] [Q&A]
What does this set of transformation do?
https://github.com/NVlabs/instant-ngp/blob/master/scripts/colmap2nerf.py#L328

Why are y and z flipped and swapped?

Why is the quaternion flipped (https://github.com/NVlabs/instant-ngp/blob/master/scripts/colmap2nerf.py#L323)? Should the transform be wordl2cam and then inverted for cam2wld transform?

What convention should the cam2wld be? Documentation on transformations and conventions used would be helpful

**Comment by Vadkoz** [comment]
I just passing by, but anyway, maybe [this](https://github.com/NVlabs/instant-ngp/discussions/463) discussion will be helpful.  

Anyway, mentioned quaternion flipping is just a code mistake because the rotation of Q is the same as the rotation of -Q due to math reasons. So maybe it's not a mistake, but rather a redundancy. 

Y and Z flipped, probably because of convention. You can project as like as on this image, but also you can do it in other direction: to the rigtht, so order of points on line become O_c, P_c, P. The for change convention you need flip Y and Z axes. I can be wrong, I really don't remember what did i do here a year ago :] Math can be tricky here. For sure documentation is highly required for such projects. 
![image](https://user-images.githubusercontent.com/41836293/231311740-acc62cc9-e8fa-4be6-8068-bdca24c89ece.png)

(i took picture from [this book](https://szeliski.org/Book/), page 55)



upd:
Indeed: 
![image](https://user-images.githubusercontent.com/41836293/231315156-59a10437-9aed-4053-ab43-560fecfc58db.png)
![image](https://user-images.githubusercontent.com/41836293/231315173-6df0f7cd-3c4a-46e3-9720-baf688124ed2.png)

Pretty usefull book. Recommend ;)
